### PR TITLE
Flag a challenge

### PR DESCRIPTION
**Pull request by tsun812**
add capability for any logged-in user to flag a challenge.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
     "@nivo/core": "^0.79.0",
     "@nivo/line": "^0.79.1",
     "@nivo/radar": "^0.79.1",
+    "@octokit/core": "^4.1.0",
```
Comment (jschwarz2030) [inline package.json:14]: can you explain why adding this library is necessary?

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 2]
```diff
@@ -66,6 +114,19 @@ export class ChallengeDetail extends Component {
       );
     }
 
+    const re = /(?!#)(\d+)\s/g
+    let currentIssues = JSON.parse(localStorage.getItem('allFlags')) || []
+    console.log(currentIssues)
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:119]: remove

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 3]
```diff
@@ -66,6 +114,19 @@ export class ChallengeDetail extends Component {
...
+    if (this.state.challengeFlagged != true) {
+      for (let i = 0; i < currentIssues.length; i++) {
+        let findMatch = currentIssues[i].title.match(re)
+        if (findMatch && findMatch[0] == challenge.id) {
+          this.setState({challengeFlagged: true, issue: currentIssues[i]})
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:124]: setting state in a render method is not a good pattern

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 4]
```diff
@@ -152,6 +213,15 @@ export class ChallengeDetail extends Component {
           ),
         });
 
+    const handleFlag = () => {
+      if (this.state.challengeFlagged) {
+          window.open(this.state.issue.html_url, "_blank")
+      } else {
+        this.setState({ ...this.state, modalToggle: true, modalClosed: true })
+      }
+    }
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:222]: this should be a handler method

File: src/components/ChallengeDetail/FlagModal.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        <h2 className="mr-text-grey-light-more mr-text-4xl mr-mt-4">Flag this challenge</h2>
+        <div className="mr-text-base mr-mt-4 mr-text-yellow">
+         <FormattedMessage {...messages.modalSubtitle} />
+        </div>
+        <FlagCommentInput challenge={props.challenge} {...props} />
+      </Modal>
+    </External>)
+}
+
+export default FlagModel
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/FlagModal.js:26]: syntax looks bad on this file.  please use prettier on new files

File: package.json (hunk 1 above)
Comment (tsun812) [reply package.json:14]: we use octokit to fetch from github api, should we use fetch api instead?

Comment (jschwarz2030) [reply package.json:14]: yes.  we should stick to the same method the rest of the repo uses.  no need to add complexity to the app.

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 6]
```diff
@@ -38,24 +40,171 @@ const ClusterMap = WithChallengeTaskClusters(
...
-    window.scrollTo(0, 0);
+  paginated_fetch = (
+    url, // Improvised required argument in JS
+    page = 1,
+    previousResponse = []
+  ) => {
+    return fetch(`${url}?state=all&sort=updated&page=${page}&per_page=30`, {
+      method: 'GET',
+      authorization: 'token ' + process.env.REACT_APP_GITHUB_ISSUES_API_TOKEN,
+      owner: 'tsun812',
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:64]: shouldn't this be a variable too?

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 7]
```diff
@@ -38,24 +40,171 @@ const ClusterMap = WithChallengeTaskClusters(
...
+      repo: 'api_test',
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:65]: and this

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 8]
```diff
@@ -38,24 +40,171 @@ const ClusterMap = WithChallengeTaskClusters(
...
+          return this.paginated_fetch(url, page, response);
+        }
+
+        return response;
+      });
   }
+  getIssues = async (id) => {
+    const currentId = parseInt(id)
+    const page = 1
+    const response = await fetch(`https://api.github.com/repos/tsun812/api_test/issues?state=all&sort=updated&page=${page}&per_page=100`, {
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:82]: and this

File: src/components/ChallengeDetail/ChallengeDetail.js [hunk 9]
```diff
@@ -38,24 +40,171 @@ const ClusterMap = WithChallengeTaskClusters(
...
+      method: 'GET',
+      authorization: 'token ' + process.env.REACT_APP_GITHUB_ISSUES_API_TOKEN,
+      owner: 'tsun812',
+      repo: 'api_test',
+    })
+    //const response = await this.paginated_fetch('https://api.github.com/repos/tsun812/api_test/issues')
```
Comment (jschwarz2030) [inline src/components/ChallengeDetail/ChallengeDetail.js:88]: remove?

**Comment by jschwarz2030** [comment]
Closing in favor of https://github.com/maproulette/maproulette3/pull/1933 due to merge conflicts

**Pull request closed without merging**